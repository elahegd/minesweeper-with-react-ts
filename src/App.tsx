import React, { FC, Suspense, lazy } from 'react';
import { Provider } from "react-redux";
import styled from "@emotion/styled";
import { BrowserRouter as Router, Route, Link, Routes, Navigate, useSearchParams } from "react-router-dom";
import { Location } from "history";

import { store } from './pages/store';

const GameWithHooks = lazy(() => import("./pages/MinesweeperWithHooks/MinesweeperWithHooks"));
const MinesweeperWithUseReducer = lazy(() => import("./pages/MinesweeperWithUseReducer/MinesweeperWithUseReducer"));
const MinesweepeerWithReactRedux = lazy(() => import("./pages/MinesweepeerWithReactRedux/MinesweepeerWithReactRedux"));

export const Navigation: FC = () => {
    const [query] = useSearchParams();
    const level = query.get('level') || "";

    const getLevelParam = (pathname: string): Partial<Location> => {
        return {
            pathname,
            search:  `${
                level &&
                `?${new URLSearchParams({
                  level,
                }).toString()}`
              }`
        }
    }

    return (
        <Nav>
            <NavigationList>
                <List><Link className="nav-link" to={getLevelParam("/")}>Home</Link></List>
                <List><Link className="nav-link" to={getLevelParam("/minesweeper/hooks")}>Game With Hooks</Link></List>
                <List><Link className="nav-link" to={getLevelParam("/minesweeper/usereducer")}>Game With Reducer</Link></List>
                <List><Link className="nav-link" to={getLevelParam("/minesweeper/react-redux")}>Game With React-redux</Link></List>
            </NavigationList>
        </Nav> 
    )
}

export const App: FC = () => (
    <Router>
        <Navigation />
        <Routes>
            <Route
                index
                element={<Home />}
            />
            <Route path="/minesweeper">    
                <Route
                    path="hooks"
                    element={
                        <Suspense fallback={<>Loading game with hooks</>}>
                            <GameWithHooks />
                        </Suspense>
                    }
                >
                    <Route 
                        path=":username?"
                        element={
                            <Suspense fallback={<>Loading game with hooks</>}>
                                <GameWithHooks />
                            </Suspense>
                        }
                    >
                    
                    </Route>
                </Route>
            </Route>
            <Route path="/minesweeper">    
                <Route
                    path="usereducer"
                    element={
                        <Suspense fallback={<>Loading game with useReducer</>}>
                            <MinesweeperWithUseReducer />
                        </Suspense>
                    }
                />
            </Route>
            <Route path="/minesweeper">
                <Route
                    path="react-redux"
                    element={
                        <Suspense fallback={<>Loading game with react-redux</>}>
                            <Provider store={store}>
                                <MinesweepeerWithReactRedux />
                            </Provider> 
                        </Suspense>
                    }
                />
            </Route>
            <Route path="*" element={<Navigate to="/" />} />
      </Routes>
    </Router>
);

export const Home: FC = () => <Text>Welcome to the Minesweeper game</Text>

const Nav = styled.div`
  background: #f5f5f5;
  display: flex;
  max-width: 100%;
  justify-content: center;
`;

const NavigationList = styled.div`
   display: flex;
   list-style: none;
   justify-content: space-between;
`;

const List = styled.li`
    padding: 10px;
    text-decoration: none;
    
`;

const Text = styled.h2`
   text-align: center;
`;