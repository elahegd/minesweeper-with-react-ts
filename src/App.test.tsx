import React from 'react';
import { render, screen } from '@testing-library/react';
import { App, Home} from './App';

const searchParams = { get: () => null };
const mockSetSearchParams = jest.fn();
jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as object),
  useSearchParams: () => [searchParams, mockSetSearchParams],
  useParams: () => ({ usename: "test" }),
}));

describe("App test case", () => {
  it("App renders check", () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("Home renders check", () => {
    const { asFragment } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();
  })
})
