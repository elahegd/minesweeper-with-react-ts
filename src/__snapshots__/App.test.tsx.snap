// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test case App renders check 1`] = `
<DocumentFragment>
  <nav>
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/minesweeper/hooks"
        >
          Game With Hooks
        </a>
      </li>
      <li>
        <a
          href="/minesweeper/usereducer"
        >
          Game With Redcer
        </a>
      </li>
      <li>
        <a
          href="/minesweeper/react-redux"
        >
          Game With React-redux
        </a>
      </li>
    </ul>
  </nav>
  <h2>
    Mineisdjf
  </h2>
</DocumentFragment>
`;

exports[`App test case Home renders check 1`] = `
<DocumentFragment>
  <h2>
    Mineisdjf
  </h2>
</DocumentFragment>
`;
